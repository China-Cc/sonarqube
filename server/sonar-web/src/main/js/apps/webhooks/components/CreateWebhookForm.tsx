/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Modal from '../../../components/controls/Modal';
import { Webhook } from '../../../app/types';
import { translate } from '../../../helpers/l10n';

interface Props {
  onClose: () => void;
  onDone: (data: { name: string; url: string }) => Promise<void>;
  webhook?: Webhook;
}

interface State {
  name: string;
  submitting: boolean;
  url: string;
}

export default class CreateWebhookForm extends React.PureComponent<Props, State> {
  mounted: boolean;

  constructor(props: Props) {
    super(props);
    const { webhook } = props;
    this.state = {
      name: webhook ? webhook.name : '',
      url: webhook ? webhook.url : '',
      submitting: false
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleCancelClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.props.onClose();
  };

  handleNameChange = (event: React.SyntheticEvent<HTMLInputElement>) =>
    this.setState({ name: event.currentTarget.value });

  handleUrlChange = (event: React.SyntheticEvent<HTMLInputElement>) =>
    this.setState({ url: event.currentTarget.value });

  handleFormSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.setState({ submitting: true });
    this.props
      .onDone({ name: this.state.name, url: this.state.url })
      .then(this.props.onClose, () => {
        if (this.mounted) {
          this.setState({ submitting: false });
        }
      });
  };

  render() {
    const { submitting } = this.state;
    const isUpdate = !!this.props.webhook;
    const modalHeader = isUpdate ? translate('webhooks.update') : translate('webhooks.create');

    return (
      <Modal contentLabel={modalHeader} onRequestClose={this.props.onClose}>
        <form onSubmit={this.handleFormSubmit}>
          <div className="modal-head">
            <h2>{modalHeader}</h2>
          </div>

          <div className="modal-body">
            <div className="modal-field">
              <label htmlFor="webhook-name">
                {translate('name')}
                <em className="mandatory">*</em>
              </label>
              <input
                autoFocus={true}
                disabled={submitting}
                id="webhook-name"
                onChange={this.handleNameChange}
                minLength={2}
                name="name"
                required={true}
                type="text"
                value={this.state.name}
              />
            </div>
            <div className="modal-field">
              <label htmlFor="webhook-url">
                {translate('webhooks.url')}
                <em className="mandatory">*</em>
              </label>
              <input
                disabled={submitting}
                id="webhook-url"
                onChange={this.handleUrlChange}
                name="url"
                required={true}
                type="text"
                value={this.state.url}
              />
              <div className="modal-field-description">{translate('webhooks.url.description')}</div>
            </div>
          </div>

          <footer className="modal-foot">
            {submitting && <i className="spinner spacer-right" />}
            <button disabled={submitting} type="submit">
              {isUpdate ? translate('update_verb') : translate('create')}
            </button>
            <button
              className="button-link"
              disabled={submitting}
              onClick={this.handleCancelClick}
              type="reset">
              {translate('cancel')}
            </button>
          </footer>
        </form>
      </Modal>
    );
  }
}
