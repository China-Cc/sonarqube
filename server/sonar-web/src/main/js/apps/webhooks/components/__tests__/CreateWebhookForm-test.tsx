/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import CreateWebhookForm from '../CreateWebhookForm';
import { submit } from '../../../../helpers/testUtils';

const webhook = { key: '1', name: 'foo', url: 'http://foo.bar' };

it('should render correctly when creating a new webhook', () => {
  expect(getWrapper()).toMatchSnapshot();
});

it('should render correctly when updating a webhook', () => {
  expect(getWrapper({ webhook })).toMatchSnapshot();
});

it('should correctly call onDone and onClose after submit', async () => {
  const onClose = jest.fn();
  const onDone = jest.fn(() => Promise.resolve());
  const wrapper = getWrapper({ onClose, onDone, webhook });
  submit(wrapper.find('form'));
  expect(wrapper).toMatchSnapshot();

  await new Promise(setImmediate);
  expect(onDone).toHaveBeenCalledWith({ name: webhook.name, url: webhook.url });
  expect(onClose).toHaveBeenCalled();
});

function getWrapper(props = {}) {
  return shallow(
    <CreateWebhookForm onClose={jest.fn()} onDone={jest.fn(() => Promise.resolve())} {...props} />
  );
}
