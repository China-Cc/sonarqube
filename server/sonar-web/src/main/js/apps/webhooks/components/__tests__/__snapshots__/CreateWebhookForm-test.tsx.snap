// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly call onDone and onClose after submit 1`] = `
<Modal
  contentLabel="webhooks.update"
  onRequestClose={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-head"
    >
      <h2>
        webhooks.update
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <div
        className="modal-field"
      >
        <label
          htmlFor="webhook-name"
        >
          name
          <em
            className="mandatory"
          >
            *
          </em>
        </label>
        <input
          autoFocus={true}
          disabled={true}
          id="webhook-name"
          minLength={2}
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
          value="foo"
        />
      </div>
      <div
        className="modal-field"
      >
        <label
          htmlFor="webhook-url"
        >
          webhooks.url
          <em
            className="mandatory"
          >
            *
          </em>
        </label>
        <input
          disabled={true}
          id="webhook-url"
          name="url"
          onChange={[Function]}
          required={true}
          type="text"
          value="http://foo.bar"
        />
        <div
          className="modal-field-description"
        >
          webhooks.url.description
        </div>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <i
        className="spinner spacer-right"
      />
      <button
        disabled={true}
        type="submit"
      >
        update_verb
      </button>
      <button
        className="button-link"
        disabled={true}
        onClick={[Function]}
        type="reset"
      >
        cancel
      </button>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly when creating a new webhook 1`] = `
<Modal
  contentLabel="webhooks.create"
  onRequestClose={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-head"
    >
      <h2>
        webhooks.create
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <div
        className="modal-field"
      >
        <label
          htmlFor="webhook-name"
        >
          name
          <em
            className="mandatory"
          >
            *
          </em>
        </label>
        <input
          autoFocus={true}
          disabled={false}
          id="webhook-name"
          minLength={2}
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="modal-field"
      >
        <label
          htmlFor="webhook-url"
        >
          webhooks.url
          <em
            className="mandatory"
          >
            *
          </em>
        </label>
        <input
          disabled={false}
          id="webhook-url"
          name="url"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <div
          className="modal-field-description"
        >
          webhooks.url.description
        </div>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <button
        disabled={false}
        type="submit"
      >
        create
      </button>
      <button
        className="button-link"
        disabled={false}
        onClick={[Function]}
        type="reset"
      >
        cancel
      </button>
    </footer>
  </form>
</Modal>
`;

exports[`should render correctly when updating a webhook 1`] = `
<Modal
  contentLabel="webhooks.update"
  onRequestClose={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-head"
    >
      <h2>
        webhooks.update
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <div
        className="modal-field"
      >
        <label
          htmlFor="webhook-name"
        >
          name
          <em
            className="mandatory"
          >
            *
          </em>
        </label>
        <input
          autoFocus={true}
          disabled={false}
          id="webhook-name"
          minLength={2}
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
          value="foo"
        />
      </div>
      <div
        className="modal-field"
      >
        <label
          htmlFor="webhook-url"
        >
          webhooks.url
          <em
            className="mandatory"
          >
            *
          </em>
        </label>
        <input
          disabled={false}
          id="webhook-url"
          name="url"
          onChange={[Function]}
          required={true}
          type="text"
          value="http://foo.bar"
        />
        <div
          className="modal-field-description"
        >
          webhooks.url.description
        </div>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <button
        disabled={false}
        type="submit"
      >
        update_verb
      </button>
      <button
        className="button-link"
        disabled={false}
        onClick={[Function]}
        type="reset"
      >
        cancel
      </button>
    </footer>
  </form>
</Modal>
`;
